### PR TITLE
re-enable script hot reload behind setreg

### PR DESCRIPTION
**Pull request by petrocket**
## What does this PR do?

re-enables Lua script hot-reloading at game time behind a setreg flag.  hot reload is a handy feature to have for rapid testing without having to exit/re-enter game mode and it lets you do a bit of development outside the editor without having to reload the level.

This feature used to exist but was removed in this commit https://github.com/o3de/o3de/commit/52ebc819804a499c61fb208cd82619dccbcb227f

## How was this PR tested?

- entered game mode and made changes to an existing lua script and verified debug statements printed in console after the updated script was saved without having to exit and re-enter game mode

### CODE REVIEW

**Review by gadams3 — APPROVED** [review]
I'm not sure why it was removed initially but this looks good to me.

File: Code/Framework/AzFramework/AzFramework/Script/ScriptComponent.cpp [hunk 1]
```diff
@@ -553,10 +565,28 @@ namespace AzFramework
...
-    //=========================================================================
-    // LoadInContext
-    // [3/3/2014]
-    //=========================================================================
+    void ScriptComponent::Deactivate()
+    {
+        if(AZ::Data::AssetBus::Handler::BusIsConnectedId(m_script.GetId()))
+        {
+            AZ::Data::AssetBus::Handler::BusDisconnect(m_script.GetId());
+        }
```
[the remark this answers is not in the document]
Comment (AMZN-alexpete) [reply Code/Framework/AzFramework/AzFramework/Script/ScriptComponent.cpp:573]: doh, need to look at my code more closely when night coding :)

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by lsemp3d — APPROVED** [review]

**Review by AMZN-daimini — APPROVED** [review]

**Comment by AMZN-alexpete** [comment]
the reason I'm holding on merging this is that in my testing, the setreg doesn't appear to work.

**Comment by petrocket** [comment]
> the reason I'm holding on merging this is that in my testing, the setreg doesn't appear to work.

Never mind, looks like I had a typo.  This .setreg works
```json
{
	"O3DE": {
		"AzFramework": {
			"Script": {
				"HotReloadEnabled": true
			}
		}
	}
}
```

There is one known issue with this hot reloading - it seems to always lag behind by one version of the file.  e.g.
1. enter game mode
2. add `Debug.Log('1')` to the script's `OnActivate` and save
3. the message does not print out
4. add `Debug.Log('2')` to the script's `OnActivate` and save
5. see '1' printed in the console
6. remove the `Debug.Log` command from `OnActivate` and save
7. see '2' printed in the console

The hot reload seems to always load the previous version of the script.  I haven't yet determined why.

**Comment by lumberyard-employee-dm** [comment]
> > the reason I'm holding on merging this is that in my testing, the setreg doesn't appear to work.
> 
> Never mind, looks like I had a typo. This .setreg works
> 
> ```json
> {
> 	"O3DE": {
> 		"AzFramework": {
> 			"Script": {
> 				"HotReloadEnabled": true
> 			}
> 		}
> 	}
> }
> ```
> 
> There is one known issue with this hot reloading - it seems to always lag behind by one version of the file. e.g.
> 
>     1. enter game mode
> 
>     2. add `Debug.Log('1')` to the script's `OnActivate` and save
> 
>     3. the message does not print out in the console as expected
> 
>     4. add `Debug.Log('2')` to the script's `OnActivate` and save
> 
>     5. see '1' printed in the console
> 
>     6. remove the `Debug.Log` command from `OnActivate` and save
> 
>     7. see '2' printed in the console
> 
> 
> The hot reload seems to always load the previous version of the script. I haven't yet determined why.

There probably needs to be a Tick frame that occurs that clears out the old script.

**Pull request merged**